docs/spec/overload.rst

```diff
@@ -99,8 +109,376 @@ variable is not sufficient.
+
+
+Step 3: If step 2 produces errors for all overloads, perform
+"argument type expansion". Some types can be decomposed
+into two or more subtypes. For example, the type ``int | str`` can be
+expanded into ``int`` and ``str``.
+
+Expansion should be performed one argument at a time from left to
+right. Each expansion results in sets of effective argument types.
+For example, if there are two arguments whose types evaluate to
+``int | str`` and ``int | bytes``, expanding the first argument type
+results in two sets of argument types: ``(int, ?)`` and ``(str, ?)``.
+Here ``?`` represents an unexpanded argument type.
+If expansion of the second argument is required, four sets of
+argument types are produced: ``(int, int)``, ``(int, bytes)``,
+``(str, int)``, and ``(str, bytes)``.
+
+After each argument expansion, return to step 2 and evaluate all
+expanded argument lists.
+
+- If all argument lists evaluate successfully, combine their
+respective return types by union to determine the final return type
+for the call, and stop.
+- If argument expansion has been applied to all arguments and one or
+more of the expanded argument lists cannot be evaluated successfully,
+generate an error and stop.
+
+For additional details about argument type expansion, see
+:ref:`argument-type-expansion` below.
+
+
+Step 4: If the argument list is compatible with two or more overloads,
+determine whether one or more of the overloads has a variadic parameter
+(either ``*args`` or ``**kwargs``) that maps to a corresponding argument
+that supplies an indeterminate number of positional or keyword arguments.
+If so, eliminate overloads that do not have a variadic parameter.
+
+- If this results in only one remaining candidate overload, it is
+the winning match. Evaluate it as if it were a non-overloaded function
+call and stop.
+- If two or more candidate overloads remain, proceed to step 5.
+
+
+Step 5: If the type of one or more arguments evaluates to a
+:ref:`gradual type` (e.g. ``list[Any]`` or ``str | Any``), determine
+whether some theoretical :ref:`materialization` of these gradual types
+could be used to disambiguate between two or more of the remaining
+overloads.
+
+- If none of the arguments evaluate to a gradual type, proceed to step 6.
+- If one or more arguments evaluate to a gradual type but no possible
+materializations of these types would disambiguate between the remaining
+overloads, proceed to step 6.
+- If possible materializations of these types would disambiguate between
+two or more of the remaining overloads and this subset of overloads have
+consistent return types, proceed to step 6. If the return types include
+type variables, constraint solving should be applied here before testing
+for consistency.
+- If none of the above conditions are met, the presence of gradual types
+leads to an ambiguous overload selection. Assume a return type of ``Any``
+and stop. This preserves the "gradual guarantee".
```

Comment:
I'm not sure this preserves the gradual guarantee as I understand it. The gradual guarantee starts with an assumption that the program being described can be coherently described and just either can't be in the current type system or hasn't been by the author. (That is, Any and other gradual types still represent a valid but unknown type if the program is consistent and sound)

If none of the above conditions are met, I believe this would not be a coherently describable function, and that the overloads are incorrect rather than imprecise.

Reply:
> If none of the above conditions are met, I believe this would not be a coherently describable function while being consistent with the overload defintions, which would mean that for the gradual guarantee to apply,we have to assume that it's a coherent program without defect, so the overloads would need to be incorrect rather than just imprecise.

I don't agree with that analysis. If you'd like to provide more details or examples that back up your analysis, please post to the typing forum thread.